### ISSUE TITLE
Label's judgment criteria ?

### ISSUE THREAD

**Issue by abc744587** [doriguzzi/lucid-ddos#7]
Dear @doriguzzi,

I am analyzing your Source code, and would like to ask how you judge **DDoS & Benign traffic** in the Label stage of 
`Traffic pre-processing`?

In the `Traffic pre-processing`  file, I see that you explained using `IP address` to Label and `5-tuple(SrcIP, SrcPort, DstIP, DstPort, Protocol)` & ` features (f=11)` mentioned in the paper, but I think What is the standard of judgment?

Thank you for listening 😃.

**Comment by doriguzzi** [comment]
Hi,

you are right, `IP addresses`  are used to determine whether a flow is benign or part of a DDoS attack. The logic for labelling the flows in the pre-processing procedure consists of comparing source and destination IP addresses of packets with the IPs in the dictionaries that are defined at the beginning of the file lucid_dataset_parser.py.
For instance, let's consider the DDoS attack in the CIC-IDS2017 dataset. The attack is generated by the host with IP address 172.16.0.1 and targets a victim with IP address 192.168.10.50, as defined in the dictionary:
`IDS2017_DDOS_FLOWS = {'attackers': ['172.16.0.1'],
                      'victims': ['192.168.10.50']}`

Therefore, every flow with packets with such IP addresses in the header are labelled with label `1' (which means DDoS), all the others are considered benign and labelled with `0'.

Hope this clarifies.

**Comment by doriguzzi** [comment]
Hi,

you are perfectly right, a 2-tuple with just source and destination IP addresses could be enough with the datasets that I used for testing. The 5-tuple has been introduced for the sake of generality, thus for more complex cases in which it is necessary a finer-grained flow classification. I mean those cases in which only part of the communication between two hosts is malicious. In such scenarios, you also need the TCP/UDP ports and the protocol information to avoid blocking regular traffic.

Said that, if you plan to use LUCID only with simple datasets, such as those from the University of New Brunswick, the IP addresses are enough.

Hope this clarifies.

**Comment by vickyngo-code** [comment]
> Hi,
> 
> you are perfectly right, a 2-tuple with just source and destination IP addresses could be enough with the datasets that I used for testing. The 5-tuple has been introduced for the sake of generality, thus for more complex cases in which it is necessary a finer-grained flow classification. I mean those cases in which only part of the communication between two hosts is malicious. In such scenarios, you also need the TCP/UDP ports and the protocol information to avoid blocking regular traffic.
> 
> Said that, if you plan to use LUCID only with simple datasets, such as those from the University of New Brunswick, the IP addresses are enough.
> 
> Hope this clarifies.

Hi @doriguzzi 

I'm actually just looking for this same answer. I'm trying to use LUCID with SWaT dataset (attacks in cyber physical systems), and it seems that just using IP addresses won't do because the packets are malformed between the two devices.

How should I initialize the 5-tuple for data labelling in lucid_data_parser.py?

**Comment by abc744587** [comment]
Dear @doriguzzi,

Thanks for your answer 😃.

I have the same question as @vickyngo-code , except IP address(Src&Dst), how are other 3-tuple used for Label?
Because if DDoS attacks and normal traffic use Port80 (HTTP) to communicate, it may cause misjudgment?
Is the LUCID method mentioned? I only mentioned it in `B. Lucid Model Architecture` in the paper, but that is the part of CNN, not pre-processing.

Above, please correct me if I understand wrong, thanks😃!

**Comment by vickyngo-code** [comment]
> Dear @doriguzzi,
> 
> Thanks for your answer 😃.
> 
> I have the same question as @vickyngo-code , except IP address(Src&Dst), how are other 3-tuple used for Label? Because if DDoS attacks and normal traffic use Port80 (HTTP) to communicate, it may cause misjudgment? Is the LUCID method mentioned? I only mentioned it in `B. Lucid Model Architecture` in the paper, but that is the part of CNN, not pre-processing.
> 
> Above, please correct me if I understand wrong, thanks😃!

Hi @abc744587,

Section `A. Network traffic preprocessing` did mention the traffic processing algorithm, and there are two parts **features extraction** and **labellings** that might be of your interest. It sorts of boils down to using flow identifier ID (e.g. DOS2017) to do the labellings.

When I examined the code, it looks like parse_labels() will eventually call process_pcap(), which will call parse_packets() and extracted 11 features as seen in Table II. It seems as if the other 3-tuple came from parse_packets().

That's why I'm confused at how are other 3-tuple are used as input for labelling, or if we are restricted with only using IP addresses. Clearly attacks such as false data injection (where packets are hijacked), do not have definite attackers or victims IP since the attacks themselves were executed from the 'inside', from my understanding.

Dear @doriguzzi 
Aside from my question on initializing the 5-tuples, I'm just wondering if there are limits on what protocols LUCID can deal with? The SWaT dataset I mentioned mostly use CIP (common industrial protocol) and possibly Modbus. They both are based on TCP, however.

I attempted to deduce the attackers & victim IP addresses of the SWaT dataset and used that to process the data, but was unsuccessful.

**Comment by doriguzzi** [comment]
> > Hi,
> > you are perfectly right, a 2-tuple with just source and destination IP addresses could be enough with the datasets that I used for testing. The 5-tuple has been introduced for the sake of generality, thus for more complex cases in which it is necessary a finer-grained flow classification. I mean those cases in which only part of the communication between two hosts is malicious. In such scenarios, you also need the TCP/UDP ports and the protocol information to avoid blocking regular traffic.
> > Said that, if you plan to use LUCID only with simple datasets, such as those from the University of New Brunswick, the IP addresses are enough.
> > Hope this clarifies.
> 
> Hi @doriguzzi
> 
> I'm actually just looking for this same answer. I'm trying to use LUCID with SWaT dataset (attacks in cyber physical systems), and it seems that just using IP addresses won't do because the packets are hijacked when they communicate between the two devices.
> 
> How should I initialize the 5-tuple for data labelling in lucid_data_parser.py?

Hi @vickyngo-code ,

in the cases in which the IP addresses are not enough, I would suggest creating a more complex structure such as
`IDS2017_DDOS_FLOWS = {'attackers': [{IP:'172.16.0.1',port:12341}], 'victims': [{IP:'192.168.10.50',port:80}]}`
Thus, a list of dictionaries, one dictionary for each attacker and each victim. I'm assuming you still deal with TCP/IP.
Of course, you also need to implement some extra checks in function `parse_labels`. But it shouldn't be too hard.

Hope this helps.

--roberto

**Comment by doriguzzi** [comment]
> Dear @doriguzzi Aside from my question on initializing the 5-tuples, I'm just wondering if there are limits on what protocols LUCID can deal with? The SWaT dataset I mentioned mostly use CIP (common industrial protocol) and possibly Modbus. They both are based on TCP, however.
> 

Dear @vickyngo-code ,

LUCID is tightly implemented around the TCP/IP protocol. Therefore it works as long as the attackers and victims can be identified with the 5-tuple of IP, TCP/UDP ports and protocol. 
In the case of other protocols, which might be implemented in the application layer, the current version of LUCID won't work, since it does not consider the content of the packets' payload.
I think you could adapt LUCID to work with other protocols if you find a dissector for TShark that supports that protocol. But, you also need to change quite deeply the method `parse_packet`, which extracts the packet's features from the TCP/IP header. In your case, you might want to extract different features that are more related to the protocol of interest.

Hope this clarifies a bit.

--roberto

**Comment by vickyngo-code** [comment]
> > Dear @doriguzzi Aside from my question on initializing the 5-tuples, I'm just wondering if there are limits on what protocols LUCID can deal with? The SWaT dataset I mentioned mostly use CIP (common industrial protocol) and possibly Modbus. They both are based on TCP, however.
> 
> Dear @vickyngo-code ,
> 
> LUCID is tightly implemented around the TCP/IP protocol. Therefore it works as long as the attackers and victims can be identified with the 5-tuple of IP, TCP/UDP ports and protocol. In the case of other protocols, which might be implemented in the application layer, the current version of LUCID won't work, since it does not consider the content of the packets' payload. I think you could adapt LUCID to work with other protocols if you find a dissector for TShark that supports that protocol. But, you also need to change quite deeply the method `parse_packet`, which extracts the packet's features from the TCP/IP header. In your case, you might want to extract different features that are more related to the protocol of interest.
> 
> Hope this clarifies a bit.
> 
> --roberto

Thank you so much for your reply. That clears things up for me. I suppose LUCID is using the signature-based detection approach? 

The live inference section shows that a `--dataset-type` must be specified. How will the model use the addressing scheme in this case? Does it have something to do with the attackers & victims IP addresses in that addressing scheme? I imagine that the model is using the attack patterns in the addressing scheme that it learned, in which case the command below will probably returns a 0% accuracy.

(Specifying a different dataset type for live inference)
`python3 lucid_cnn.py --predict_live eth0 --model ./sample-dataset/10t-10n-DOS2019-LUCID.h5 --dataset_type SYN2020`

**Comment by doriguzzi** [comment]
> > > Dear @doriguzzi Aside from my question on initializing the 5-tuples, I'm just wondering if there are limits on what protocols LUCID can deal with? The SWaT dataset I mentioned mostly use CIP (common industrial protocol) and possibly Modbus. They both are based on TCP, however.
> > 
> > 
> > Dear @vickyngo-code ,
> > LUCID is tightly implemented around the TCP/IP protocol. Therefore it works as long as the attackers and victims can be identified with the 5-tuple of IP, TCP/UDP ports and protocol. In the case of other protocols, which might be implemented in the application layer, the current version of LUCID won't work, since it does not consider the content of the packets' payload. I think you could adapt LUCID to work with other protocols if you find a dissector for TShark that supports that protocol. But, you also need to change quite deeply the method `parse_packet`, which extracts the packet's features from the TCP/IP header. In your case, you might want to extract different features that are more related to the protocol of interest.
> > Hope this clarifies a bit.
> > --roberto
> 
> Thank you so much for your reply. That clears things up for me. I suppose LUCID is using the signature-based detection approach?
> 
> The live inference section shows that a `--dataset-type` must be specified. How will the model use the addressing scheme in this case? Does it have something to do with the attackers & victims IP addresses in that addressing scheme? I imagine that the model is using the attack patterns in the addressing scheme that it learned, in which case the command below will probably returns a 0% accuracy.
> 
> (Specifying a different dataset type for live inference) `python3 lucid_cnn.py --predict_live eth0 --model ./sample-dataset/10t-10n-DOS2019-LUCID.h5 --dataset_type SYN2020`

Hi,

let me first state that option `--dataset-type` is not mandatory for inference. If present, LUCID checks the attackers/victim addressing scheme specified in the dictionaries defined in the parser to print the classification statistics (accuracy, F1, TNR, etc.). If the dataset type is not specified, such statistics are not printed. 

 `--dataset-type` has been introduced only for testing reasons, to help the developer to test the system in online settings, for instance replaying one of the pcap files of one of the UNB datasets with `tcpreplay`. In the wild, this option is obviously not usable, as LUCID would print wrong statistics (even though the classification would not be affected).

In summary,   `--dataset-type` should be used to verify the accuracy of LUCID in online settings with the `pcap` files of the dataset used for training/validation/testing in offline settings. If the accuracy is very different from that obtained with the test set, there might be some problems in the transmission of the pcap file over the network (packet fragmentation, mismatch of MAC addresses, etc.)
`--dataset-type` should not be used in the wild, with unknown traffic, as it will cause LUCID to print wrong results (although the detection accuracy is not affected).

Hope this answer your questions.

--roberto
